### ISSUE TITLE
Improvements to ad load behavior

### ISSUE THREAD

**Issue by cramforce** [ampproject/amphtml#4022]
Currently we consider an ad loaded when the iframe loaded. For some networks (such as DFP) we can observe when actual ad render start and potentially keep showing the fallback until this time.

Similarly we can expose a promise from `amp-ad` that resolves when an ad was rendered and whether the slot was filled. This could be used by `amp-sticky-ad` to only ever show the banner at this time.

CC @jasti

**Comment by jasti** [comment]
@cramforce two things.
1. "for showing fallback until this time" - since a fallback can be basically anything, including "We have no ads to show for you" - I don't know if we can do this by default.

2. I like the promise idea - do ad networks have to implement some some sort of ad render start callback to invoke to trigger this promise?

**Comment by cramforce** [comment]
Not all ad network do this. But we can make it work only where we know it is supported. Your new Team will help to announce this being supported.

**Comment by zhouyx** [comment]
#4124

**Comment by lannka** [comment]
Seems work is done.
For the follow up work that to make use of this event, we can open new issue to track.

**Comment by lannka** [comment]
We're making some changes to the APIs.

Right now, `render-start` will always expected to be called, either by 3P code or `integration.js`. After that, `no-content` might be called by 3P code.

This is bad because:
- `no-content` after `render-start` will trigger layout twice.
- Runtime has no way to differentiate if the `render-start` is called by `integration.js` which means only the bootstrap iframe is loaded, or by 3P code which means the ad is ready. It causes trouble for us to correctly measure ads loading latency.

The changes are:
- Differentiate `render-start` called by `integration.js` and 3P.  `integration.js` will call `bootstrapped` (naming TBD) instead. And it'll be always called.
- Runtime will listen to `render-start` and `no-content` only once. Both APIs are supposed to be called by 3P. And the 2 events will race. Runtime will only handle whichever arrives first. 
- Runtime maintains a list of 3Ps that has implemented `render-start` API. For those 3P, runtime will change visibility base on `render-start`, otherwise base on `bootstrapped`.

@zhouyx

**Comment by zhouyx** [comment]
`render-start` API is re-implemented. It provides ways for multi-size ads to let AMP runtime resize. #3176. It also provides info about ad status to AMP runtime.

**Comment by jasti** [comment]
Closing this issue since render-start is implemented and publicly announced to devs.
